File: 12_end/01_ingest_traffic.R
# 01_ingest_brussels_traffic_counts_realtime.R
# Ingest Brussels Realtime Traffic Counts (1m, t1)
# Tim Fraser
#
# This cron-friendly script fetches the latest traverse-level vehicle counts from
# the Brussels traffic API and stores normalized rows in SQLite.

# Run from inside the 12_end/ directory so the paths resolve correctly.
# Git bash: cd 12_end && Rscript 01_ingest_brussels_traffic_counts_realtime.R
# Powershell: Set-Location 12_end; Rscript 01_ingest_brussels_traffic_counts_realtime.R

# 0. SETUP ###################################

## 0.1 Load Packages #################################

library(httr2)
library(dplyr)
library(tibble)
library(purrr)
library(lubridate)
library(DBI)
library(RSQLite)

if (file.exists(".env")) readRenviron(".env")

# 1. CONFIG ###################################

# Brussels Traffic Vehicle Counts API Documentation here:
# https://data.mobility.brussels/traffic/api/counts/
BASE_URL = "https://data.mobility.brussels/traffic/api/counts/"
BRUSSELS_METRO_ID = 948
script_arg = commandArgs(trailingOnly = FALSE)
script_path = sub("^--file=", "", script_arg[grepl("^--file=", script_arg)][1])
script_dir = if (!is.na(script_path) && nzchar(script_path)) {
  dirname(normalizePath(script_path, winslash = "/", mustWork = FALSE))
} else {
  normalizePath("12_end", winslash = "/", mustWork = FALSE)
}
DATA_DIR = file.path(script_dir, "data")
DB_PATH = file.path(DATA_DIR, "traffic.db")

dir.create(DATA_DIR, showWarnings = FALSE, recursive = TRUE)

cat("\n====================================================\n")
cat("📋 01_ingest_brussels_traffic_counts_realtime.R\n")
cat("====================================================\n")
cat("   🧭 metro_id: ", BRUSSELS_METRO_ID, "\n", sep = "")
cat("   🔗 ", BASE_URL, "\n", sep = "")

# 2. FETCH DATA ###################################

resp = request(BASE_URL) |>
  req_url_query(request = "live", includeLanes = "false", interval = "1") |>
  req_perform()

if (resp_status(resp) != 200) {
  stop("Brussels traffic API failed with status ", resp_status(resp))
}

body = resp_body_json(resp, simplifyVector = FALSE)
if (is.null(body$data) || length(body$data) == 0) {
  stop("Brussels traffic API returned empty data payload")
}

monitors = names(body$data)
cat("   ✅ monitors in payload: ", length(monitors), "\n", sep = "")

# 3. CLEAN DATA ###################################

# Write a function to extract the data for each monitor
extract = function(monitor_id, monitor_payload) {
  one_min = monitor_payload$results[["1m"]]
  t1 = if (is.null(one_min) || is.null(one_min$t1)) list() else one_min$t1
  vehicles = if (is.null(t1$count)) NA else t1$count
  speed = if (is.null(t1$speed)) NA else t1$speed
  occupancy = if (is.null(t1$occupancy)) NA else t1$occupancy
  end_time_raw = if (is.null(t1$end_time)) "" else as.character(t1$end_time)

  tibble(
    monitor_id = monitor_id,
    vehicles = suppressWarnings(as.integer(vehicles)),
    speed = suppressWarnings(as.numeric(speed)),
    occupancy = suppressWarnings(as.numeric(occupancy)),
    observed_at_raw = end_time_raw
  )
}

# For each monitor, extract that monitor's data and bind as dataframe
raw_df = map2_dfr(
  .x = monitors, .y = body$data,
  .f = ~extract(monitor_id = .x, monitor_payload = .y) )

# Convert Brussels timestamp to a valid UTC timestamp
parse_bxl_time = function(x) {
  parsed = suppressWarnings(ymd_hm(x, tz = "Europe/Brussels", quiet = TRUE))
  ifelse(is.na(parsed), NA_character_, format(with_tz(parsed, "UTC"), "%Y-%m-%d %H:%M:%S"))
}

# Clean the data!
df = raw_df |>
  mutate(observed_at = parse_bxl_time(observed_at_raw)) |>
  filter(
    !is.na(vehicles),
    !is.na(observed_at),
    !is.na(monitor_id),
    monitor_id != ""
  ) |>
  mutate(speed = if_else(speed < 0, true = 0, false = speed)) |>
  transmute(
    metro_id = BRUSSELS_METRO_ID,
    monitor_id = monitor_id,
    observed_at = observed_at,
    vehicles = vehicles,
    speed = speed,
    occupancy = occupancy
  )

if (nrow(df) == 0) {
  stop("No valid 1m/t1 monitor rows parsed from Brussels API payload")
}

cat("   ✅ parsed rows: ", nrow(df), "\n", sep = "")
print(slice_head(df, n = 3))

# 4. WRITE TO SQLITE ###################################
# Keep database logic intentionally minimal and easy to read.

# Connect to/initialize the database
db = dbConnect(SQLite(), DB_PATH)

# Create the table if it doesn't exist
invisible(dbExecute(db, "
  CREATE TABLE IF NOT EXISTS traffic (
    metro_id    INTEGER,
    monitor_id  TEXT,
    observed_at TEXT,
    vehicles    INTEGER,
    speed       REAL,
    occupancy   REAL,
    PRIMARY KEY (metro_id, monitor_id, observed_at)
  )
"))

# Create a unique index on the table (will prevent duplicates)
invisible(dbExecute(
  db,
  "
  CREATE UNIQUE INDEX IF NOT EXISTS idx_traffic_metro_monitor_observed
  ON traffic (metro_id, monitor_id, observed_at)
"
))

before_count = dbGetQuery(
  db,
  "
  SELECT COUNT(*) AS n
  FROM traffic
  WHERE metro_id = :metro_id
",
  params = list(metro_id = BRUSSELS_METRO_ID)
)$n[[1]]

# Insert the data into the table
invisible(dbExecute(
  db,
  "
  INSERT INTO traffic (metro_id, monitor_id, observed_at, vehicles, speed, occupancy)
  VALUES (:metro_id, :monitor_id, :observed_at, :vehicles, :speed, :occupancy)
  ON CONFLICT(metro_id, monitor_id, observed_at) DO NOTHING
",
  params = df
))

# Check total rows
total_rows = dbGetQuery(
  db,
  "
  SELECT COUNT(*) AS n
  FROM traffic
  WHERE metro_id = :metro_id
",
  params = list(metro_id = BRUSSELS_METRO_ID)
)$n[[1]]
inserted_rows = max(total_rows - before_count, 0)

# Disconnect from the database
dbDisconnect(db)

# 5. VERIFY ###################################
cat("   ✅ candidate rows this run: ", nrow(df), "\n", sep = "")
cat("   ✅ new rows appended: ", inserted_rows, "\n", sep = "")
cat("   ✅ total rows (metro): ", total_rows, "\n", sep = "")


# That's it! You've made a script that you can run manually or via CRON 
# to ingest the latest Brussels traffic counts into a SQLite database.

File: 12_end/01_ingest_traffic.py
# 01_ingest_traffic.py
# Ingest Brussels Realtime Traffic Counts (1m, t1)
# Pairs with 01_ingest_traffic.R
# Tim Fraser
#
# This cron-friendly script fetches the latest traverse-level vehicle counts from
# the Brussels traffic API and stores normalized rows in SQLite.

# Run from inside the 12_end/ directory so the paths resolve correctly.
# Git bash: cd 12_end && python 01_ingest_traffic.py
# Powershell: Set-Location 12_end; python 01_ingest_traffic.py

# 0. SETUP ###################################

## 0.1 Load Packages #################################

import sqlite3
import time
from datetime import datetime
from pathlib import Path
from zoneinfo import ZoneInfo

import requests


# 1. CONFIG ###################################

# Brussels Traffic Vehicle Counts API documentation:
# https://data.mobility.brussels/traffic/api/counts/
BASE_URL = "https://data.mobility.brussels/traffic/api/counts/"
BRUSSELS_METRO_ID = 948
SCRIPT_DIR = Path(__file__).resolve().parent
DATA_DIR = SCRIPT_DIR / "data"
DB_PATH = DATA_DIR / "traffic.db"
DATA_DIR.mkdir(parents=True, exist_ok=True)

print("\n====================================================")
print("01_ingest_traffic.py | Brussels realtime ingest")
print("====================================================")
print(f"   metro_id: {BRUSSELS_METRO_ID}")
print(f"   api: {BASE_URL}")


def get_with_retry(url: str, params: dict, max_attempts: int = 5, timeout: int = 30) -> requests.Response:
    """Fetch API payload with retry/backoff for transient failures."""
    for attempt in range(1, max_attempts + 1):
        response = requests.get(url, params=params, timeout=timeout)
        if response.status_code in {429, 500, 502, 503, 504} and attempt < max_attempts:
            retry_after = response.headers.get("Retry-After")
            sleep_seconds = int(retry_after) if retry_after and retry_after.isdigit() else min(2 ** attempt, 30)
            print(
                f"   warning transient status={response.status_code}; "
                f"retrying in {sleep_seconds}s (attempt {attempt}/{max_attempts})"
            )
            time.sleep(sleep_seconds)
            continue
        response.raise_for_status()
        return response
    raise RuntimeError("Failed to fetch traffic payload after retries.")


def parse_bxl_time_to_utc(end_time: str) -> str | None:
    """Convert Brussels-local timestamp to UTC string format used in SQLite."""
    if not end_time:
        return None
    for fmt in ("%Y-%m-%d %H:%M", "%Y/%m/%d %H:%M"):
        try:
            local_dt = datetime.strptime(end_time, fmt).replace(tzinfo=ZoneInfo("Europe/Brussels"))
            utc_dt = local_dt.astimezone(ZoneInfo("UTC"))
            return utc_dt.strftime("%Y-%m-%d %H:%M:%S")
        except (TypeError, ValueError):
            continue
    return None


# 2. FETCH DATA ###################################

# Ask the API for the "live" request payload at one-minute interval granularity.
response = get_with_retry(
    BASE_URL,
    params={"request": "live", "includeLanes": "false", "interval": "1"},
    timeout=30,
)
payload = response.json()
data = payload.get("data", {})

# Hard-fail fast if API returns an empty payload so the cron run is visibly red.
if not data:
    raise SystemExit("Brussels traffic API returned empty data payload.")

monitors = list(data.keys())
print(f"   monitors in payload: {len(monitors)}")


# 3. CLEAN DATA ###################################

# Convert nested monitor payloads into flat rows with expected schema and types.
rows = []
for monitor_id, monitor_payload in data.items():
    one_min = (monitor_payload.get("results", {}) or {}).get("1m", {}) or {}
    t1 = one_min.get("t1", {}) or {}
    vehicles = t1.get("count")
    speed = t1.get("speed")
    occupancy = t1.get("occupancy")
    observed_at_raw = t1.get("end_time", "")
    observed_at = parse_bxl_time_to_utc(observed_at_raw)

    # Skip malformed rows early to keep downstream SQL simple and robust.
    if vehicles is None or observed_at is None or not monitor_id:
        continue

    try:
        row = (
            BRUSSELS_METRO_ID,
            str(monitor_id),
            observed_at,
            int(vehicles),
            max(float(speed), 0.0) if speed is not None else None,
            float(occupancy) if occupancy is not None else None,
        )
        rows.append(row)
    except (TypeError, ValueError):
        continue

if not rows:
    raise SystemExit("No valid 1m/t1 monitor rows parsed from Brussels API payload.")

print(f"   parsed rows: {len(rows)}")
print(f"   sample row: {rows[0]}")


# 4. WRITE TO SQLITE ###################################

# Keep database logic intentionally minimal and easy to read for students.
conn = sqlite3.connect(str(DB_PATH))
conn.execute(
    """
    CREATE TABLE IF NOT EXISTS traffic (
      metro_id    INTEGER,
      monitor_id  TEXT,
      observed_at TEXT,
      vehicles    INTEGER,
      speed       REAL,
      occupancy   REAL,
      PRIMARY KEY (metro_id, monitor_id, observed_at)
    )
"""
)
conn.execute(
    """
    CREATE UNIQUE INDEX IF NOT EXISTS idx_traffic_metro_monitor_observed
    ON traffic (metro_id, monitor_id, observed_at)
"""
)

# Capture count before insert so we can report new rows appended this run.
before_count = conn.execute(
    "SELECT COUNT(*) FROM traffic WHERE metro_id = ?",
    (BRUSSELS_METRO_ID,),
).fetchone()[0]

# Insert rows and ignore duplicates so repeated cron runs stay idempotent.
conn.executemany(
    """
    INSERT INTO traffic (metro_id, monitor_id, observed_at, vehicles, speed, occupancy)
    VALUES (?, ?, ?, ?, ?, ?)
    ON CONFLICT(metro_id, monitor_id, observed_at) DO NOTHING
""",
    rows,
)
conn.commit()

# Count rows after insert and calculate net-new inserts for this run.
total_rows = conn.execute(
    "SELECT COUNT(*) FROM traffic WHERE metro_id = ?",
    (BRUSSELS_METRO_ID,),
).fetchone()[0]
inserted_rows = max(int(total_rows) - int(before_count), 0)
conn.close()


# 5. VERIFY ###################################

print(f"   candidate rows this run: {len(rows)}")
print(f"   new rows appended: {inserted_rows}")
print(f"   total rows (metro): {int(total_rows)}")
